hal/spi_api.h

```diff
@@ -19,220 +18,305 @@
+ * - Calling any function other than `spi_init()` after calling `spi_free()`.
+ * - Calling `spi_async_free_handle()` more than one time on a `spi_async_handle_t`.
+ * - Calling `spi_async_abort()` after calling `spi_async_free_handle()`.
+ *
+ * # What this API does not cover
+ * The following elements are not covered by this API and are considered implementation details :
+ * - The use of Interrupts and/or DMA for async operations.
+ * - The way `spi_transfer()` is implemented : using specific to this function or using
+ *   `spi_async_transfer()`.
+ * - Wether the SS pin is controlled by hardware or software.
+ * - The way `spi_async_abort()` is implemented : abortion callback invoked in the same thread before
+ *   returning, in an interrupt or in another thread.
+ *
+ * @{
+ *
+ * \struct spi_async_handle_t
+ * This needs to be declared and defined by the low level device driver.
+ * It is used to eventually abort an async request before its completion. see spi_async_abort.
+ */
 
-#define SPI_EVENT_ERROR       (1 << 1)
-#define SPI_EVENT_COMPLETE    (1 << 2)
-#define SPI_EVENT_RX_OVERFLOW (1 << 3)
-#define SPI_EVENT_ALL         (SPI_EVENT_ERROR | SPI_EVENT_COMPLETE | SPI_EVENT_RX_OVERFLOW)
-
-#define SPI_EVENT_INTERNAL_TRANSFER_COMPLETE (1 << 30) // Internal flag to report that an event occurred
-
-#define SPI_FILL_WORD         (0xFFFF)
-#define SPI_FILL_CHAR         (0xFF)
-
-#if DEVICE_SPI_ASYNCH
-/** Asynch SPI HAL structure
+/**
+ * This enumerates the possible transmission mode of a SPI peripheral.
+ * The SPI_MODE_QUAD_IO mode can sometimes also be backed by a QSPI peripheral in "legacy SPI" mode.
+ *
+ * @warning Do not confuse with clock_polarity and clock_phase in spi_init_t.
  */
-typedef struct {
-    struct spi_s spi;        /**< Target specific SPI structure */
-    struct buffer_s tx_buff; /**< Tx buffer */
-    struct buffer_s rx_buff; /**< Rx buffer */
-} spi_t;
+typedef enum spi_mode_t {
+    SPI_MODE_SIMPLEX, /**< Unidirectionnal communication on a single wire. */
+    SPI_MODE_HALF_DUPLEX, /**< Bidirectionnal communication on a single wire. */
+    SPI_MODE_FULL_DUPLEX, /**< Bidrectionnal communication on two wire (MISO/MOSI). */
+    SPI_MODE_DUAL_IO, /**< Half-duplex communication on a two wire. */
+    SPI_MODE_QUAD_IO, /**< Half-duplex communication on a four wire. */
+} spi_mode_t;
 
-#else
-/** Non-asynch SPI HAL structure
+/**
+ * SPI object.
+ * The actual definition of this structure is delegated to the device implementation of the hal.
  */
 typedef struct spi_s spi_t;
 
-#endif
-
-#ifdef __cplusplus
-extern "C" {
-#endif
-
 /**
- * \defgroup hal_GeneralSPI SPI Configuration Functions
- * @{
+ * This structure groups all initialization parameters required by a SPI interface.
  */
+typedef struct spi_init_s {
+    bool    is_master;  /**< True to configure the device in Master mode */
+    bool    msb_first;  /**< True to send/receive the most significant bit first. */
```

Comment:
This being a C-library, you should add `#include <stdbool.h>`, right?

Reply:
As well as `#include <stdint.h>` for `uint32_t` and `#include "device.h"` for `PinName`. Good catch !